mirrord/layer/src/tcp.rs

```diff
@@ -82,13 +85,41 @@ pub(crate) trait TcpHandler {
         }
     }
 
+    /// Handle a potential [`LayerError::NewConnectionAfterSocketClose`] error.
+    fn check_connection_handling_result(res: Result<(), LayerError>) -> Result<(), LayerError> {
+        if let Err(LayerError::NewConnectionAfterSocketClose(connection_id)) = res {
+            // This can happen and is valid:
+            // 1. User app closes socket.
+            // 2. mirrord sends `PortUnsubscribe` to agent.
+            // 3. Agent sniffs new incoming connection and sends it to the layer.
+            // 4. Agent receives `PortUnsubscribe`.
+            // Step 2 could even happen after 4.
+            if Self::IS_STEAL {
+                warn!(
+                    "Got incoming tcp connection (mirrord connection id: {connection_id}) after \
+                    the application already closed the socket. The connection will not be handled \
+                    and will be reset once the agent is informed about the application closing the \
+                    socket."
+                )
+            } else {
+                info!(
+                    "Got incoming tcp connection (mirrord connection id: {connection_id}) after \
+                    the application already closed the socket."
+                );
+            }
+            Ok(())
+        } else {
+            res
+        }
+    }
+
     /// Returns true to let caller know to keep running
     #[tracing::instrument(level = "trace", skip(self))]
     async fn handle_daemon_message(&mut self, message: DaemonTcp) -> Result<(), LayerError> {
         let handled = match message {
-            DaemonTcp::NewConnection(tcp_connection) => {
-                self.handle_new_connection(tcp_connection).await
-            }
+            DaemonTcp::NewConnection(tcp_connection) => Self::check_connection_handling_result(
+                self.handle_new_connection(tcp_connection).await,
+            ),
```

Comment:
```suggestion
            DaemonTcp::NewConnection(tcp_connection) => self.handle_new_connection(tcp_connection).await.and_then(Self::check_connection_handling_result),
```

Looks nicer? Less nice? Think they're equivalent right?

Reply:
Here:
```
Self::check_connection_handling_result(
	                self.handle_new_connection(tcp_connection).await,
	            )
```
`check_connection_handling_result` is called on the `Result`  of `handle_new_connection`. With `and_then` it would only be called on an `Ok`.

Reply:
Oh! Yup, ... errmm ... hmm huh... it was a ... huh ... trick question! Yes, a trick question! And you have passed, congratulations ....

(sorry, my brain created a sort of `chain` method for `Result` I guess)